[tosa] Add legalization for TF/TFL conv3d

tatwaichong opened the pull request:
TFL doesn't have quantized implementation support for conv3d at the moment, only floating is supported in this patch.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -68,32 +68,28 @@ LogicalResult ConvertTFBiasAddOp::matchAndRewrite(
 
   auto value = tf_biasadd_op.value();
   auto bias = tf_biasadd_op.bias();
+  auto bias_shape = bias.getType().dyn_cast<RankedTensorType>().getShape();
```
Just cast unless you check if null first before using.

tatwaichong replied:
Got you. Thanks for letting me know.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -68,32 +68,28 @@ LogicalResult ConvertTFBiasAddOp::matchAndRewrite(
   auto bias = tf_biasadd_op.bias();
+  auto bias_shape = bias.getType().dyn_cast<RankedTensorType>().getShape();
+  // Bias tensor that feeds into tosa.conv2d or tosa.conv3d must be rank 1
+  if (bias_shape.size() != 1) return failure();
```
Could you use notifyMatchFailure here too?

tatwaichong replied:
Sure thing. I use this manner throughout the newer patch as well.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -68,32 +68,28 @@ LogicalResult ConvertTFBiasAddOp::matchAndRewrite(
-
+  if (tf_conv2d_op) {
   // Sanity check to confirm rhs() has the expected shape of bias
   auto filter_shape =
```
Same re cast

tatwaichong replied:
Done. s/dyn_cast/cast

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -68,32 +68,28 @@ LogicalResult ConvertTFBiasAddOp::matchAndRewrite(
-    return failure();
-  }
-
+  if (tf_conv2d_op) {
```
Nit: to reduce scope/life range/avoid accidental dereferencing when null, prefer

if (TF::Conv2DOp tf_conv2d_op = ...) { ... }

tatwaichong replied:
Fair enough. Done.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -105,6 +101,35 @@ LogicalResult ConvertTFBiasAddOp::matchAndRewrite(
+
+    if (!result) return failure();
+
+    rewriter.replaceOp(op, {result.getValue()});
```
value() - getValue has been marked as deprecated.

tatwaichong replied:
Good to know.  Changed.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -3391,6 +3391,106 @@ llvm::Optional<Value> convertTFConv2DCommon(
+      dilation = rewriter.getI64ArrayAttr({dilation_d, dilation_h, dilation_w});
+    }
+
+   return convertConv3DCommon(
```
Did clang-format produce this?

tatwaichong replied:
Sorry. I didn't run clang-format before this submission. In the next patch, I run clang-format-10 on it.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:
```diff
@@ -844,6 +845,35 @@ LogicalResult ConvertTFConv2DOp::matchAndRewrite(
+      tf_conv3d_op.getResult().getType().dyn_cast<RankedTensorType>();
+
+  // Set up a zero attr for subsequent pattern replacement if required
+  auto bias_dim = filter_type.getShape().back();
```
Missing check if ranked?

tatwaichong replied:
Good spotting. A ranked check is added.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -1089,6 +1091,91 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
+  if ((input_is_qtype != filter_is_qtype) ||
+      (input_is_qtype != output_is_qtype)) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLConv3DOp: input/filter/output tensor should "
```
Update to remove ConvertTFLConv3DOp and trailing punctuation?

tatwaichong replied:
I didn't notice MLIR error convention at that time. Removed, will never again.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -1089,6 +1091,91 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
+
+  Value unquantized_bias = tfl_conv3d_op.bias();
+  if (!unquantized_bias.getType().dyn_cast<RankedTensorType>()) {
+  // The bias may actually be typed "None" which has no value. TOSA requires
```
Missing indent?

tatwaichong committed:
[tosa] Add legalization for TF/TFL conv3d

Only support floating for now as TFL doesn't have quantized
implementation support at the moment.

tatwaichong commented:
Adopt the suggested review comments throughout the patch.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -68,46 +68,70 @@ struct ConvertTFBiasAddOp : public RewritePattern {
-  if (!output_type) return failure();
+
+  if (!output_type) {
+    return rewriter.notifyMatchFailure(op, "output not a ranksed tensor");
```
ranked

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:
```diff
@@ -68,46 +68,70 @@ struct ConvertTFBiasAddOp : public RewritePattern {
-  TF::Conv2DOp tf_conv2d_op =
-      dyn_cast_or_null<TF::Conv2DOp>(value.getDefiningOp());
+  if (TF::Conv2DOp tf_conv2d_op =
+          dyn_cast_or_null<TF::Conv2DOp>(value.getDefiningOp())) {
```
dyn_cast_if_present is the new name, we could either do this here or in follow up over entire file

tatwaichong replied:
Happy to know this. I changed to use `dyn_cast_if_present`  along with a typo above in the file.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -3396,6 +3396,102 @@ llvm::Optional<Value> convertTFConv2DCommon(
+    // Defaults to [1, 1, 1].
+    stride = rewriter.getI64ArrayAttr({1, 1, 1});
+  } else {
+    int64_t stride_d = strides_attr[1].cast<IntegerAttr>().getInt();
```
There is a probably a way to do this with getAsValueRange followed by map, but given these probably end up being converted to DenseArrayAttr instead in follow up (which would allow for nicer accessor) that's fine.

tatwaichong replied:
I've tried `getAsValueRange` but it seems that the returning `llvm::iterator_range` doesn't support random access [] operator, There is only the index 1, 2 and 3, i.e. hwd, are needed. Maybe an extra code is required to handle unnecessary elements.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -1092,6 +1094,93 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
+  int64_t stride_d = tfl_conv3d_op.stride_d();
+  int64_t stride_h = tfl_conv3d_op.stride_h();
+  int64_t stride_w = tfl_conv3d_op.stride_w();
+  ArrayAttr stride = rewriter.getI64ArrayAttr({stride_d, stride_h, stride_w});
```
This feel off: we are creating an ArrayAttr here, just to unpack and repack into an ArrayAttr in convertConv3DCommon. Creating Attributes goes through process of uniquing and is stored lifetime of context. Could these just be passed in as ArrayRef?

tatwaichong replied:
I haven't noticed the advantages of using ArrayRef before. Is the primary purpose of this to avoid copying objects? These attributes are packed in SmallVector now and passed in as ArrayRef, please check if they match your suggestion. As `getPaddingValuesFromPadType` and `CreateOpAndInfer` requires ArrayAttr as parameter, ArrayRef have to be converted to ArrayAttr when calling them. We can change getPaddingValuesFromPadType to use ArrayRef in follow up.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -1092,6 +1094,93 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
+
+  if (!a1_conv3d_op) return failure();
+
+  Value conv3d_output;
```
How does the equivalent ternary op formulation look here?

tatwaichong replied:
Look nice.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -1092,6 +1094,93 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
+  }
+
+  auto fused_activation_fn = tfl_conv3d_op.fused_activation_functionAttr();
+  if (fused_activation_fn) {
```
To reduce scope/lifeness, 

if (auto fused_activation_fn = ...) { ... }

tatwaichong replied:
Done.

tatwaichong commented:
@stellaraccident, I accidentally removed request for you.

gbaned commented:
@tatwaichong  Can you please resolve conflicts? Thank you!

tatwaichong commented:
FYI. [This](https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/fd62905d807b5976bea28b6d766e614c076faacf) is the change in the reference model to verify the legalization in TF/TFL.

jpienaar commented on tensorflow/compiler/mlir/tosa/tests/tf-to-tosa-pipeline.mlir:
```diff
@@ -42,6 +42,35 @@ func.func @test_transpose_conv2d(%arg0: tensor<1x32x32x8xf32>, %arg1: tensor<1x1
+// CHECK-SAME: %[[VAL_0:.*]]: tensor<2x4x128x128x8xf32>
+// CHECK-SAME: %[[VAL_1:.*]]: tensor<2x3x3x2x4xf32>
+// CHECK: %[[VAL_2:.*]] = "tosa.const"() {value = dense<0.000000e+00> : tensor<4xf32>}
+// CHECK: %[[VAL_3:.*]] = "tosa.const"() {value = dense<[4, 0, 1, 2, 3]> : tensor<5xi32>}
```
For all the constants, could you use CHECK-DAG instead of CHECK? (flagged in internal test which I don't know why it isn't upstream, but it is meant to not be sensitive to orderings of sequence of constants)

tatwaichong replied:
Sorry for the inconvenience. Done.
